scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala

```diff
@@ -1907,30 +1910,121 @@ class FormatOps(
     }
   }
 
+  def getOptionalBracesSplits(ft: FormatToken, ob: OptionalBraces)(implicit
+      style: ScalafmtConfig
+  ): Seq[Split] = {
+    val expire = lastToken(ob.tree)
+    def getIndent(len: Int) = Indent(Num(len), expire, ExpiresOn.After)
+    val nlPolicy = decideNewlinesOnlyAfterClose(Split(Newline, 0))(expire)
+    if (ft.hasBlankLine) {
+      val indent = getIndent(style.indent.getSignificant)
+      Seq(Split(Newline2x, 0).withIndent(indent).withPolicy(nlPolicy))
+    } else if (ob.forceNL) {
+      val indent = getIndent(style.indent.getSignificant)
+      Seq(Split(Newline, 0).withIndent(indent).withPolicy(nlPolicy))
+    } else {
+      val indentLen =
+        if (ob.mainIndent) style.indent.main else style.indent.getSignificant
+      Seq(
+        Split(Space, 0).withSingleLine(expire),
+        Split(Newline, 1).withIndent(getIndent(indentLen)).withPolicy(nlPolicy)
+      )
+    }
+  }
+
   // Optional braces in templates after `:|with`
   object OptionalBracesTemplate {
-    def unapply(meta: FormatToken.Meta): Option[Template] =
+    def unapply(meta: FormatToken.Meta): Option[OptionalBraces] =
       meta.leftOwner match {
         case t: Template if templateCurly(t).contains(tokens(meta.idx).left) =>
-          Some(t)
+          Some(new OptionalBraces(t, forceNL = true))
         case _ => None
       }
   }
 
   // Optional braces after any token that can start indentation:
-  // )  =  =>  ?=>  <-  catch  do  else  finally  for
-  // if  match  return  then  throw  try  while  yield
+  // )  =>  ?=>  <-  do  else  finally  for
+  // if  return  then  throw  try  while  yield
   object OptionalBracesBlock {
-    def unapply(meta: FormatToken.Meta): Option[Term.Block] = {
-      val nft = nextNonComment(tokens(meta.idx))
+    def unapply(meta: FormatToken.Meta): Option[OptionalBraces] = {
+      val nft = nextNonComment(meta)
```

Comment:
Seems we can't use `rightOwner` :/ for example:
```scala
def printTextDocument(doc: TextDocument): String =
      if cond then
        sb.append(doc.text.substring(offset, end))
        sb.append(doc.text.substring(offset, end))
```

rightOwner will actually not be Block.

Reply:
let me see

Reply:
Something like:
```
      val nft = nextNonComment(meta)
      val foundBlock = nft.meta.leftOwner match {
        case t: Term.If if meta.left.text == "then" => Some(t.thenp)
        case t: Term.If if meta.left.text == "else" => Some(t.elsep)
        case _ => Nil
      }
      foundBlock match {
        case Some(t: Term.Block) if t.tokens.headOption.contains(nft.right) =>
          Some(new OptionalBraces(t, mainIndent = t.stats.length == 1))
        case _ => None
      }
```
seems to work, but we need to extract block from each leftOwner.

Reply:
I started doing some tests on the Dotty codebase and this popped up.

Reply:
Getting to:
```
val foundBlock: Option[Term] = nft.meta.leftOwner match {
        case t: Term.If if meta.left.text == "then" || meta.left.text == ")" => Some(t.thenp)
        case t: Term.If if meta.left.text == "else" => Some(t.elsep)
        case t: Term.Try if meta.left.text == "try" => Some(t.expr)
        case t: Term.Try if meta.left.text == "finally" => t.finallyp
        case t: Term.Return => Some(t.expr)
        case t: Term.While if meta.left.text == "while"  =>  Some(t.expr)
        case t: Term.While if meta.left.text == "do"  =>  Some(t.body)
        case t: Term.If if meta.left.text == "then" || meta.left.text == ")" => Some(t.thenp)
        case _ => None
      }
```

Is there any easier way? :sweat_smile:

Reply:
#2413

somehow i feel that a lot of the complexity with scalafmt is because scalameta trees are a bit top-down and randomly structured. most expressions look to me like an infix; they have an op, a left and a right.

Reply:
> ```
> val foundBlock: Option[Term] = nft.meta.leftOwner match {
>         case t: Term.If if meta.left.text == "then" || meta.left.text == ")" => Some(t.thenp)
>         case t: Term.If if meta.left.text == "else" => Some(t.elsep)
> ...
> ```

Sorry, didn't see that earlier. I'd probably use `.is[KwThen]` or `RightParen` and avoid `text` comparisons, but i like that it doesn't traverse any trees. the only question, is this list exhaustive? lambdas, partial functions, `do` coming from `for`... let me think.

Reply:
It's not yet exhaustive, it started getting too too long, so I actually started looking at a similar approach to what you did 😅

But maybe it's just a couple of cases more.